File: ios/StripePayments.m
#import "StripePayments.h"
#import <React/RCTLog.h>
#import <React/RCTConvert.h>

@implementation StripePayments

RCT_EXPORT_MODULE()

RCT_EXPORT_METHOD(init:(NSString *)publishableKey)
{
    [StripeAPI setDefaultPublishableKey:publishableKey];
}

RCT_EXPORT_METHOD(confirmPayment:(NSString *)publishableKey secret:(NSString *)secret cardParams:(NSDictionary *)cardParams resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    [STPAPIClient sharedClient].publishableKey = publishableKey;
    // Collect card details
    STPPaymentMethodCardParams *card = [[STPPaymentMethodCardParams alloc] init];
    card.number = [RCTConvert NSString:cardParams[@"number"]];
    card.expYear = [RCTConvert NSNumber:cardParams[@"expYear"]];
    card.expMonth = [RCTConvert NSNumber:cardParams[@"expMonth"]];
    card.cvc = [RCTConvert NSString:cardParams[@"cvc"]];
    STPPaymentMethodParams *paymentMethodParams = [STPPaymentMethodParams paramsWithCard:card billingDetails:nil metadata:nil];
    STPPaymentIntentParams *paymentIntentParams = [[STPPaymentIntentParams alloc] initWithClientSecret:secret];
    paymentIntentParams.paymentMethodParams = paymentMethodParams;
    paymentIntentParams.setupFutureUsage = @(STPPaymentIntentSetupFutureUsageOnSession);

    // Submit the payment
    STPPaymentHandler *paymentHandler = [STPPaymentHandler sharedHandler];
    [paymentHandler confirmPayment:paymentIntentParams withAuthenticationContext:self completion:^(STPPaymentHandlerActionStatus status, STPPaymentIntent *paymentIntent, NSError *error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            switch (status) {
                case STPPaymentHandlerActionStatusFailed: {
                    if ([error.debugDescription rangeOfString:@"You cannot confirm this PaymentIntent because it has already succeeded after being previously confirmed."].location == NSNotFound) {
                        reject(@"StripeModule.failed", error.localizedDescription, nil);
                    } else {
                        reject(@"StripeModule.failed", @"You cannot confirm this PaymentIntent because it has already succeeded after being previously confirmed.", nil);
                    }
                    break;
                }
                case STPPaymentHandlerActionStatusCanceled: {
                    reject(@"StripeModule.cancelled", @"", nil);
                    break;
                }
                case STPPaymentHandlerActionStatusSucceeded: {
                    resolve(@{
                        @"id": paymentIntent.allResponseFields[@"id"],
                        @"paymentMethodId": paymentIntent.paymentMethodId
                    });
                    break;
                }
                default:
                    break;
            }
        });
    }];
}

- (UIViewController *)authenticationPresentingViewController
{
    return RCTPresentedViewController();
}

@end

File: ios/StripePayments.swift
import Foundation
import Stripe

@objc(StripePayments)
class StripePayments: NSObject {

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc(init:)
    func `init`(publishableKey: String) {
        STPAPIClient.shared.publishableKey = publishableKey
    }

    @objc(isCardValid:)
    func isCardValid(cardParams: NSDictionary) -> Bool {
        return true
    }

    @objc(confirmPayment:cardParams:resolver:rejecter:)
    func confirmPayment(
        secret: String,
        cardParams: [String: Any],
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        let card = STPPaymentMethodCardParams()
        card.number = RCTConvert.nsString(cardParams["number"])
        card.expYear = RCTConvert.nsNumber(cardParams["expYear"])
        card.expMonth = RCTConvert.nsNumber(cardParams["expMonth"])
        card.cvc = RCTConvert.nsString(cardParams["cvc"])

        let methodParams = STPPaymentMethodParams(card: card, billingDetails: nil, metadata: nil)
        let intentParams = STPPaymentIntentParams(clientSecret: secret)
        intentParams.paymentMethodParams = methodParams
        intentParams.setupFutureUsage = .onSession

        STPPaymentHandler.shared().confirmPayment(intentParams, with: self) { status, paymentIntent, error in
            switch status {
            case .succeeded:
                var result: [String: Any] = [:]
                result["id"] = paymentIntent?.allResponseFields["id"]
                result["paymentMethodId"] = paymentIntent?.paymentMethodId
                resolver(result)
            case .canceled:
                rejecter("StripeModule.cancelled", "", nil)
            case .failed:
                rejecter("StripeModule.failed", error?.localizedDescription ?? "", nil)
            }
        }
    }
}

extension StripePayments: STPAuthenticationContext {
    func authenticationPresentingViewController() -> UIViewController {
        RCTPresentedViewController()!
    }
}
